crates/llm-local/src/token_output_stream.rs

```diff
@@ -0,0 +1,68 @@
+/// Implementation for TokenOutputStream Code is borrowed from
+/// https://github.com/huggingface/candle/blob/main/candle-examples/src/token_output_stream.rs
+/// (Commit SHA 4fd00b890036ef67391a9cc03f896247d0a75711)
+///
+/// This is a wrapper around a tokenizer to ensure that tokens can be returned to the user in a
+/// streaming way rather than having to wait for the full decoding.
+pub struct TokenOutputStream {
+    tokenizer: tokenizers::Tokenizer,
+    tokens: Vec<u32>,
+    prev_index: usize,
+    current_index: usize,
+}
+
+impl TokenOutputStream {
+    pub fn new(tokenizer: tokenizers::Tokenizer) -> Self {
+        Self {
+            tokenizer,
+            tokens: Vec::new(),
+            prev_index: 0,
+            current_index: 0,
+        }
+    }
+
+    fn decode(&self, tokens: &[u32]) -> anyhow::Result<String> {
+        match self.tokenizer.decode(tokens, true) {
+            Ok(str) => Ok(str),
+            Err(err) => anyhow::bail!("cannot decode: {err}"),
+        }
```

Comment:
```suggestion
       self.tokenizer.decode(tokens, true).context("failed to decode token stream")
```

Reply:
It does look like I cannot do this because `tokenizer.decode` returns a  `Result<String, Box<dyn Error + Send + Sync>>` which does not seem to be suitable to use context on(?)